### ISSUE TITLE
Direct3D API interception and modification

### ISSUE THREAD

**Issue by copslock** [elishacloud/dxwrapper#52]
Hello Elisha!
It's great lucky to find some code of Direct3D API logging but as you told this layer can be used to modify the D3D API calls from game program,so I'm thinking about solving an historical problem that annoying me for years----the Call of Duty 4(iw3 engine) on Intel GMA Gen3(900/950/3100/3150) GPUs.
The problem is the game will failed with
`Create2DTexture( $floatz, 640, 480, 0, 114 ) failed: 8876086c = Invalid call`
By analyzing the API calls of iw3sp.exe,I found that after the creation of D3D window,the program send such command
`IDirect3DDevice9_CreateTexture, IDirect3DDevice9P<11a208a8>, UINT<280>, UINT<1e0>, UINT<1>, DWORD<1>, D3DFORMAT<**D3DFMT_R32F**>, D3DPOOL<D3DPOOL_DEFAULT>, IDirect3DTexture9PP<1d43614>, HANDLEP<0>, Format<72>, D3DFORMAT<**D3DFMT_R32F**>no POW2 texture!!!
no POW2 texture!!!
, HRESULT<8876086c>`
As the code above,the program send rendering command using `D3DFMT_R32F` format,and according to the header file of swiftshader(https://github.com/bkaradzic/SwiftShader/blob/master/src/D3D9/Direct3D9.cpp)
```
case D3DRTYPE_SURFACE:
			if(usage & D3DUSAGE_RENDERTARGET)
			{
				switch(checkFormat)
				{
				case D3DFMT_NULL:			if(!Capabilities::Surface::RenderTarget::NULL_)			return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_R8G8B8:			if(!Capabilities::Surface::RenderTarget::R8G8B8)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_R5G6B5:			if(!Capabilities::Surface::RenderTarget::R5G6B5)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_X1R5G5B5:		if(!Capabilities::Surface::RenderTarget::X1R5G5B5)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A1R5G5B5:		if(!Capabilities::Surface::RenderTarget::A1R5G5B5)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A4R4G4B4:		if(!Capabilities::Surface::RenderTarget::A4R4G4B4)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_R3G3B2:			if(!Capabilities::Surface::RenderTarget::R3G3B2)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A8R3G3B2:		if(!Capabilities::Surface::RenderTarget::A8R3G3B2)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_X4R4G4B4:		if(!Capabilities::Surface::RenderTarget::X4R4G4B4)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A8R8G8B8:		if(!Capabilities::Surface::RenderTarget::A8R8G8B8)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_X8R8G8B8:		if(!Capabilities::Surface::RenderTarget::X8R8G8B8)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A8B8G8R8:		if(!Capabilities::Surface::RenderTarget::A8B8G8R8)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_X8B8G8R8:		if(!Capabilities::Surface::RenderTarget::X8B8G8R8)		return NOTAVAILABLE();	else return D3D_OK;
				// Integer HDR formats
				case D3DFMT_G16R16:			if(!Capabilities::Surface::RenderTarget::G16R16)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A2B10G10R10:	if(!Capabilities::Surface::RenderTarget::A2B10G10R10)	return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A2R10G10B10:	if(!Capabilities::Surface::RenderTarget::A2R10G10B10)	return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A16B16G16R16:	if(!Capabilities::Surface::RenderTarget::A16B16G16R16)	return NOTAVAILABLE();	else return D3D_OK;
				// **Floating-point formats**
				case D3DFMT_R16F:			if(!Capabilities::Surface::RenderTarget::R16F)			return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_G16R16F:		if(!Capabilities::Surface::RenderTarget::G16R16F)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A16B16G16R16F:	if(!Capabilities::Surface::RenderTarget::A16B16G16R16F)	return NOTAVAILABLE();	else return D3D_OK;
				case **D3DFMT_R32F**:			if(!Capabilities::Surface::RenderTarget::R32F)			return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_G32R32F:		if(!Capabilities::Surface::RenderTarget::G32R32F)		return NOTAVAILABLE();	else return D3D_OK;
				case D3DFMT_A32B32G32R32F:	if(!Capabilities::Surface::RenderTarget::A32B32G32R32F)	return NOTAVAILABLE();	else return D3D_OK;
				default:
					return NOTAVAILABLE();
				}
			}
```
The `D3DFMT_R32F` render target is actually Floating-point render target format,such format is avaliable even on the Geforce FX5200,but nor the Intel.
In the develop guide of GMA900(Same Gen3 core of GMA3100) P16
https://software.intel.com/sites/default/files/m/d/4/1/d/8/Intel_915G_SDG_Feb05.pdf
And the introduction of Intel X3000 Unified gpu architecture P13
https://www.intel.com/Assets/PDF/whitepaper/313343.pdf
Intel said it only supports fix-point render target and do not support the floating-type on Gen3 architecture.
So everything becomes clear,it's here any possible to modify the HRESULT of that call to SUCCESS or much better converting it into equally fix-point type instruction?
Some guys may talk with me that the Intel GMA3100 is completely crap and don't waste time on such rubbish but as you know in the old days of 2007-era even such crappy GPU gives much more pleasure than today's TitanX,but at that time I wasn't know very much about programming and it's quite regretful at that time.It's 10 years later and I'm finally be able to find what is going on with that game.Hope you can give a little advice to solve this problem 
Best regards

**Comment by VladWinner** [comment]
> Create2DTexture( $floatz, 640, 480, 0, 114 ) failed: 8876086c = Invalid call
> Call of Duty 4(iw3 engine)

One of the users informed me about similar error in Call of Duty 2 (DX9) on Intel GPU:
<img width="573" height="96" alt="Image" src="https://github.com/user-attachments/assets/bd3f166e-bfe7-4ac6-98ee-7f8816ce278e" />
<img width="210" height="65" alt="Image" src="https://github.com/user-attachments/assets/ea18c499-83ae-4a59-8401-ce76bbc062f3" />

Here are the properties of mentioned texture in XnView:
<img width="471" height="407" alt="Image" src="https://github.com/user-attachments/assets/c0884915-b004-474f-a4c0-e3a3a7fa332c" />
(BC3 RGBA 8 bpp DXT5 is a fairly standard format for a 2005 game)

@elishacloud Can dxwrapper theoretically know in advance whether the GPU supports these formats and issue a warning before the game starts its first attempts at graphics initialization? I'm not sure what exactly the problem is and how to identify incompatible computers in order to suggest these users to change the wrapper.

**Comment by elishacloud** [comment]
> Can dxwrapper theoretically know in advance whether the GPU supports these formats and issue a warning before the game starts its first attempts at graphics initialization?

Yes, that could be done.  You can create a d3d9 object and query for D3DFMT_R32F support.  But I wouldn't want to do that kind of custom changes in dxwrapper.  I prefer to put a fix in to make the game work even if the device doesn't support D3DFMT_R32F.

**Comment by Squall-Leonhart** [comment]
there poc gpu's don't even have a proper vertex shader.
